src/dbnode/storage/index.go

```diff
@@ -1335,6 +1354,45 @@ func (i *nsIndex) Query(
 	}, nil
 }
 
+func (i *nsIndex) WideQuery(
+	ctx context.Context,
+	query index.Query,
+	collector chan *ident.IDBatch,
+	opts index.WideQueryOptions,
+) error {
+	logFields := []opentracinglog.Field{
+		opentracinglog.String("wideQuery", query.String()),
+		opentracinglog.String("namespace", i.nsMetadata.ID().String()),
+		opentracinglog.Int("batchSize", opts.BatchSize),
+		xopentracing.Time("queryStart", opts.StartInclusive),
+		xopentracing.Time("queryEnd", opts.EndExclusive),
+	}
+
+	ctx, sp := ctx.StartTraceSpan(tracepoint.NSIdxWideQuery)
+	sp.LogFields(logFields...)
+	defer sp.Finish()
+
+	results := index.NewWideQueryResults(
+		i.nsMetadata.ID(),
+		i.opts.IdentifierPool(),
+		i.shardForID(),
+		collector,
+		opts,
+	)
+
+	ctx.RegisterFinalizer(results)
+	queryOpts := opts.ToQueryOptions()
+
+	defer results.Finalize()
```

Comment:
Seems to be finalizing twice, once at the end of this func and once with `ctx.RegsiterFinalizer(...)`.

Do you want to remove the `ctx.RegisterFinalizer(results)` if you truly don't need results at the end?

Reply:
Talked: Remove cloning of IDs in result and use `ctx.RegisterFinalizer(...)` and then remove `defer results.Finalize()`.

Reply:
Refactored this a bit, added an explicit Close() method to wide results to avoid the deadlocking we talked about (the `<-doneCh` in `database.go`)